timemachine/cpp/src/neighborlist.cu

```diff
@@ -113,80 +125,136 @@ Neighborlist<RealType>::get_nblist_host(int N, const double *h_coords, const dou
+    dim3 dimGrid(row_blocks, Y, 1); // block x, y, z dims
 
     // (ytz): TBD shared memory, stream
-    k_find_blocks_with_ixns<RealType><<<dimGrid, tpb, 0, stream>>>(
-        N,
-        d_block_bounds_ctr_,
-        d_block_bounds_ext_,
-        d_coords,
-        d_box,
-        d_ixn_count_,
-        d_ixn_tiles_,
-        d_ixn_atoms_,
-        d_trim_atoms_,
-        cutoff);
+    if (!this->compute_full_matrix_) {
+        // Compute only the upper triangle as rows and cols are the same
+        // pass duplicates of column coords and the bounding boxes
+        k_find_blocks_with_ixns<RealType, true><<<dimGrid, tpb, 0, stream>>>(
+            N,
+            N,
+            d_col_block_bounds_ctr_,
+            d_col_block_bounds_ext_,
+            d_col_block_bounds_ctr_,
+            d_col_block_bounds_ext_,
+            d_col_coords,
+            d_col_coords,
+            d_box,
+            d_ixn_count_,
+            d_ixn_tiles_,
+            d_ixn_atoms_,
+            d_trim_atoms_,
+            cutoff);
+    } else {
+        k_find_blocks_with_ixns<RealType, false><<<dimGrid, tpb, 0, stream>>>(
+            N,
+            K,
+            d_col_block_bounds_ctr_,
+            d_col_block_bounds_ext_,
+            d_row_block_bounds_ctr_,
+            d_row_block_bounds_ext_,
+            d_col_coords,
+            d_row_coords,
+            d_box,
+            d_ixn_count_,
+            d_ixn_tiles_,
+            d_ixn_atoms_,
+            d_trim_atoms_,
+            cutoff);
+    }
 
     gpuErrchk(cudaPeekAtLastError());
-
-    k_compact_trim_atoms<<<B, tpb, 0, stream>>>(N, Y, d_trim_atoms_, d_ixn_count_, d_ixn_tiles_, d_ixn_atoms_);
+    k_compact_trim_atoms<<<row_blocks, tpb, 0, stream>>>(N, Y, d_trim_atoms_, d_ixn_count_, d_ixn_tiles_, d_ixn_atoms_);
 
     gpuErrchk(cudaPeekAtLastError());
 }
 
+template <typename RealType>
+void Neighborlist<RealType>::build_nblist_device(
+    const int N, const double *d_coords, const double *d_box, const double cutoff, cudaStream_t stream) {
+
+    this->build_nblist_device(N, 0, d_coords, nullptr, d_box, cutoff, stream);
+}
+
 template <typename RealType>
 void Neighborlist<RealType>::compute_block_bounds_device(
-    const int N,            // Number of atoms
-    const int D,            // Box dimensions
-    const double *d_coords, // [N*3]
-    const double *d_box,    // [D*3]
+    const int N,                // Number of atoms in column
+    const int K,                // Number of atoms in row
+    const int D,                // Box dimensions
+    const double *d_col_coords, // [N*3]
+    const double *d_row_coords, // [K*3]
+    const double *d_box,        // [D*3]
     cudaStream_t stream) {
 
-    assert(N == N_);
-    assert(D == 3);
+    if (N != N_) {
+        throw std::runtime_error("N != N_");
+    }
+    if (K != K_) {
+        throw std::runtime_error("K != K_");
+    }
+    if (D != 3) {
+        throw std::runtime_error("D != 3");
+    }
 
```

Comment:
Can we add a consistency check that if K==0 then col_ptr is null, and if K>0, then col_ptr is not null?